activerecord/lib/active_record/relation/merger.rb

```diff
@@ -81,6 +83,14 @@ def merge
       end
 
       private
+        def merge_unscope_values
+          return if other.unscope_values.empty?
+
+          if other.table == relation.table
+            relation.unscope!(*other.unscope_values)
+          end
```

Comment:
The doc says:

> Returns an array representing the intersection of the resulting records

So I think we shouldn't apply any `unscope` to the other relation, no?

Reply:
> The citation is for when other is an array.

Yes, but you'd think it's trying to behave more or less the same in both cases.

I need to go do some git history spelunking, but at first sight it seems to me that the intent of `.merge` is to product the intersection of the results.

Reply:
Arf, it was introduced in https://github.com/rails/rails/commit/a8b10a2a8d6f8d861453803264b9ef1b0cadbc6b, but no associated PR.

But we can see it's `aliased` as `&`.

So IMO, we should always strive to ensure that `a.merge(b)` behave as if we did `a.to_a & b.to_a` (which in some cases might be impossible)

Reply:
This is done in this PR. We merge `unscope`s only on relations for the same model (because for different models `a.to_a & b.to_a` makes no sense).

If changing for `unscope` to not apply to the outer relation (basically only removing it from `NORMAL_VALUES` array without any other changes), a bunch of tests start to fail (related to default scopes and `unscope`s):
```
Failure:
HasManyThroughAssociationsTest#test_has_many_through_unscope_default_scope [test/cases/associations/has_many_through_associations_test.rb:1455]:
Expected: 2
  Actual: 1

Failure:
HasAndBelongsToManyAssociationsTest#test_habtm_scope_can_unscope [test/cases/associations/has_and_belongs_to_many_associations_test.rb:949]:
Expected: 1
  Actual: 0

Failure:
HasManyAssociationsTest#test_can_unscope_the_default_scope_of_the_associated_model [test/cases/associations/has_many_associations_test.rb:2828]:
--- expected
+++ actual
@@ -1 +1 @@
-[#<Bulb ID: 8, car_id: 10, name: "defaulty", frickinawesome: false, color: nil>, #<Bulb ID: 9, car_id: 10, name: "other", frickinawesome: false, color: nil>]
+[#<Bulb ID: 8, car_id: 10, name: "defaulty", frickinawesome: false, color: nil>]

Failure:
HasManyAssociationsTest#test_unscopes_the_default_scope_of_associated_model_when_used_with_include [test/cases/associations/has_many_associations_test.rb:2856]:
--- expected
+++ actual
@@ -1 +1 @@
-[#<Bulb ID: 27, car_id: 22, name: "defaulty", frickinawesome: false, color: nil>, #<Bulb ID: 28, car_id: 22, name: "other", frickinawesome: false, color: nil>]
+[#<Bulb ID: 27, car_id: 22, name: "defaulty", frickinawesome: false, color: nil>]

Failure:
DefaultScopingTest#test_unscope_merging [test/cases/scoping/default_scoping_test.rb:488]:
Expected #<ActiveRecord::Relation::WhereClause:0x000000011a238e60 @predicates=[#<Arel::Nodes::Equality:0x000000011a23a148 @left=#<struct Arel::Attributes::Attribute relation=#<Arel::Table:0x0000000115ac47a0 @name="developers", @klass=Developer(id: integer, name: string, salary: integer, firm_id: integer, mentor_id: integer, legacy_created_at: datetime, legacy_updated_at: datetime, legacy_created_on: datetime, legacy_updated_on: datetime, last_name: ), @type_caster=#<ActiveRecord::TypeCaster::Map:0x00000001162c2060 @klass=Developer(id: integer, name: string, salary: integer, firm_id: integer, mentor_id: integer, legacy_created_at: datetime, legacy_updated_at: datetime, legacy_created_on: datetime, legacy_updated_on: datetime, last_name: )>, @table_alias=nil>, name="name">, @right=#<ActiveRecord::Relation::QueryAttribute:0x0000000119ff81c8 @name="name", @value_before_type_cast="Jamis", @type=#<ActiveModel::Type::String:0x0000000112671ae8 @true="t", @false="f", @precision=nil, @scale=nil, @limit=nil>, @original_attribute=nil>>]> to be empty.
```

If you have a better idea how to fix it, feel free to open another PR or push changes into this.

Reply:
Yeah, I'm not saying I have a better idea, just that the `merge` expected behavior isn't super clear to me.

And that's a risky change, because in the past changing this sort of behavior has created pretty big bugs in user apps. So I think we should thread very carefully here.